charts/meteor-pipelines/templates/validate-jupyterhub-image.yaml

```diff
@@ -0,0 +1,55 @@
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: validate-jupyterhub-image
+spec:
+  params:
+   - name: url
+  results:
+   - name: python-runtime
+     description: Python runtime
+   - name: pip-packages
+     description: Packages installed in runtime
+   - name: message
+     description: Messages passed down from task
+  steps:
+    - name: jupyter-book-build
+      image: $(params.url)
+      securityContext:
+        privileged: true
+      script: |
+         #!/usr/bin/env bash
+
+         # Default values for easier handling when error occurs
+         echo -n "" > $(results.message.path)
+         echo -n "[]" > $(results.python-runtime.path)
+         echo -n "[]" > $(results.pip-packages.path)
+
+         # Check Python version and save it to variable
+         echo  "Checking Python version"
+         version=$(python --version)
+         # Check if Python was found, if not fail task
+         if [[ $? != 0 ]]; then
+              echo "Python is not installed" >&2
+              echo "Python unavailable" > $(results.message.path)
+              exit 1
+         fi
+
+         # Create JSON from Python version
+         PYTHON_VERSION=$(echo "[{\"name\":\"Python\",\"version\":\"$( echo $version | cut -d ' ' -f 2)\"}]")
+
+         # Save existing Python packages
+         echo "Checking for existing Python packages"
+         PACKAGES=$(pip list --disable-pip-version-check --format json)
+        
+         if [[ $? != 0 ]]; then
```

Comment:
Same as above - let's do `command -v pip` first, instead of checking exit status.

Reply:
Actually, let's rather ask than reinvent stuff....

@fridex, @codificat, @pacospace do you know about any simple and native way to fetch the explicitly installed in Python packages only? Something that would skip all the nested dependencies so we can get a clean list. E.g. do not show `Werkzeug` if `flask` is installed.

I would like to avoid installing additional dependencies here if possible, so we don't compromise/modify the analyzed stack ([pip-chill](https://github.com/rbanffy/pip-chill) comes to mind). Also we need to be sure it would work with any possible python version since we should be able to analyze any imaginable stack.

Reply:
I was thinking that `--not-required` option in `pip list` might do the trick.  According to documentation https://pip.pypa.io/en/stable/cli/pip_list/ it is used to list packages that are not dependencies of installed packages. And from my testing (picture below) it seems to do exactly what we want unless there is some catch that I didn't notice. WDYT @tumido  
![image](https://user-images.githubusercontent.com/47832967/152109374-a37639f2-a830-41e8-a319-29f68b9626a4.png)

Reply:
The problem with this solution is that we cannot identify the index from where the package has been installed (just something to consider, even though we are not worrying about index yet for this first iteration). Moreover it seems you basically want to discard the transitive dependencies right? Should we ask today? maybe the UI can hide the transitive ones unless the admin wants to see them? wdyt?

Reply:
@pacospace We should definitely discuss this. I don't think you wan't to display transitive deps at all. The same for indexes. Mind this list is used by users on the spawner. I think they are interested in explicit deps only. The argument is - do you want to see this:
![image](https://user-images.githubusercontent.com/7453394/152111963-d8fc26a5-af76-4814-b379-6657b8813fac.png)

Or full pip list from the same package:
![image](https://user-images.githubusercontent.com/7453394/152112358-60a76bde-b3de-4584-aa9f-1edddfc66ad9.png)

It's a very different story if you want to show user list of **159** packages compared to **2** important ones currently displayed for the same image in the UI.

The `--not-required` flag looks like a great find @mimotej :+1: That way the list of 159 packages shrinks to 10. Which I think is a good starting point:
![image](https://user-images.githubusercontent.com/7453394/152112596-786423ae-ce8d-436b-9ebd-1323aa1d5978.png)

Reply:
> maybe the UI can hide the transitive ones unless the admin wants to see them?

Do you know a way how to make `pip list` output a tree structure as JSON? If you want to support such usecase, we need to find a way how to get the information what is a transitive dependency from pip. Maybe we can store this as 2 separate annotations? 1 for the full list of packages and another one for the explicitly installed? Mind this solution still doesn't represent package relation, because `pip list` doesn't output such info.

Reply:
> > maybe the UI can hide the transitive ones unless the admin wants to see them?
> 
> Do you know a way how to make `pip list` output a tree structure as JSON? If you want to support such usecase, we need to find a way how to get the information what is a transitive dependency from pip. Maybe we can store this as 2 separate annotations? 1 for the full list of packages and another one for the explicitly installed? Mind this solution still doesn't represent package relation, because `pip list` doesn't output such info.

@fridex will know more, but not sure if there is some command like `pipenv graph` or `thamos graph` without need for another package to be installed.

Reply:
@mimotej `--not-required` is in `pip` since v9.0.0 (released Nov 2016) https://pip.pypa.io/en/stable/news/#v9-0-0, I think this should be safe to use :slightly_smiling_face:

Reply:
> @mimotej `--not-required` is in `pip` since v9.0.0 (released Nov 2016) https://pip.pypa.io/en/stable/news/#v9-0-0, I think this should be safe to use slightly_smiling_face

we can do a diff between base command and this command maybe? so we don't use other packages?

Reply:
@pacospace not sure if `pipenv` or `thamos` is the right answer. We can't assume either is installed and our validation should not fail due to them being unavailable. :slightly_frowning_face: Is there a binary/python version independent version of either libs that we can simply mount to the image or something?

Reply:
We have this thread https://discuss.python.org/t/pip-installation-reports/12316 and https://github.com/thoth-station/micropipenv/issues/206 I think this will be sooner-or-later implemented upstream. I hadn't time to push any implementation, but it looks like a feature that is desperately lacking in the Python community.

Reply:
> E.g. do not show `Werkzeug` if `flask` is installed.

Are these images built by AICoE-CI? AICoE-CI puts information about direct dependencies to `/opt`, you can eventually read that file. When it comes to pip, I don't know about any option that could be helpful in this case (if I understand your case correctly).